commit fe6f212ce12341df18ef9b890bea739b4547157b
Author: Reinette Chatre <reinette.chatre@intel.com>
Date:   Mon Apr 19 10:46:31 2010 -0700

    mac80211: pass HT changes to driver when off channel
    
    Since "mac80211: make off-channel work generic" drivers have not been
    notified of configuration changes after association or authentication. This
    caused more dependence on current state to ensure driver will be notified
    when configuration changes occur. One such problem arises if off-channel is
    in progress when HT information changes. Since HT is only enabled on the
    "oper_channel" the driver will never be notified of this change. Usually
    the driver is notified soon after of a BSS information change
    (BSS_CHANGED_HT) ... but since the driver did not get a notification that
    this is a HT channel the new BSS information does not make sense.
    
    Fix this by also changing the off-channel information when HT is enabled
    and thus cause driver to be notified correctly.
    
    This fixes a problem in 4965 when associated with 5GHz 40MHz channel.
    Without this patch the system can associate but is unable to transfer any
    data, not even ping.
    
    See http://bugzilla.intellinuxwireless.org/show_bug.cgi?id=2158
    
    Signed-off-by: Reinette Chatre <reinette.chatre@intel.com>
    Signed-off-by: John W. Linville <linville@tuxdriver.com>

diff --git a/net/mac80211/mlme.c b/net/mac80211/mlme.c
index be5f723d643a..8a9650343f26 100644
--- a/net/mac80211/mlme.c
+++ b/net/mac80211/mlme.c
@@ -167,6 +167,8 @@ static u32 ieee80211_enable_ht(struct ieee80211_sub_if_data *sdata,
 	ht_changed = conf_is_ht(&local->hw.conf) != enable_ht ||
 		     channel_type != local->hw.conf.channel_type;
 
+	if (local->tmp_channel)
+		local->tmp_channel_type = channel_type;
 	local->oper_channel_type = channel_type;
 
 	if (ht_changed) {